### PR TITLE
Proposal for main database schema which contains user and streamer data

### PR DESCRIPTION
**Pull request by bausano**

### CODE REVIEW

File: 001-MAIN_SCHEMA.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
+This document presents a database schema for the first version of the project.
+The fields were gathered by browsing through competitive UIs and narrowing them down to the most important bits.
+Not all database tables are listed here.
+For example, the schema for chat is in a separate document.
+
+For large images we store both `url` and [`blurhash`][blurhash] strings.
+[Blurhash][blurhash] generates 20-30 character strings that are used to display a blurred version of the image while the actual image is loading.
+
+# `users`
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:9]: are we gonna do 2fa? feels like people are used to it with modern apps, adds to credibility and it's not too difficult to set up

File: 001-MAIN_SCHEMA.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  - value is a JSON in format `{"col1":"html to render", "col2": "...", "col3": "..."}`
+- `thumbnail`
+  - the streamer can set a thumbnail for their stream
+  - value is a JSON in format `{"url": "...", "blurhash": "..."}`
+- `preview`
+  - if the thumbnail is not set, we default to a preview
+  - the preview image can also be shown when user hovers over the streamer
+  - value is a JSON in format `{"url": "...", "blurhash": "..."}`
+
+# `streamers`
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:57]: `streamers` or `streamer_settings`? It's just semantics but any user could be treated as a streamer? Also I think we're missing some important fields here such as `stream_key`, `keep_vods`, `enable_clips`. Or if we treat `streamers as` a separate resource, we could use a similar `streamer_kv` table to store these settings.

File: 001-MAIN_SCHEMA.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+- `streamer_id` (fk, index)
+- `user_id` (fk, index)
+- `streak` (int)
+  - number of consecutive months the user has subscribed
+- `cumulative` (int)
+  - number of months the user has subscribed in total
+  - ie. how many times have we charged the user
+- `created_at` (timestamp)
+  - when the user subscribed for the first time
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:136]: `last_subscribed_at` maybe

File: 001-MAIN_SCHEMA.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+- `v` (text)
+
+User keys:
+
+- `profile_banner`
+  - value is a JSON in format `{"url": "...", "blurhash": "..."}`
+- `twitter_url`
+- `ig_url`
+- `yt_url`
+- `discord_url`
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:42]: isn't this all streamer-only stuff too?

File: 001-MAIN_SCHEMA.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+- `streamer_id` (fk, index)
+- `user_id` (fk, index)
+- `streak` (int)
+  - number of consecutive months the user has subscribed
+- `cumulative` (int)
+  - number of months the user has subscribed in total
+  - ie. how many times have we charged the user
+- `created_at` (timestamp)
+  - when the user subscribed for the first time
+
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:137]: `acitivity_log` table maybe? Sounds useful down the line. Title changes, category changes, account related stuff maybe, channel follows, subscriptions, raids etc.

File: 001-MAIN_SCHEMA.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+- `twitter_url`
+- `ig_url`
+- `yt_url`
+- `discord_url`
+
+Streamer-only keys:
+
+- `about`
+  - value used in streamer profile page
+  - value is a JSON in format `{"col1":"html to render", "col2": "...", "col3": "..."}`
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:48]: I'd be up to make this a bit more sophisticated than just "html to render"

File: 001-MAIN_SCHEMA.md (hunk 1 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:9]: Ah good point, 2FA does sound like a must. Haven't implemented it as of yet, going to do some research on it and come back in this thread.

File: 001-MAIN_SCHEMA.md (hunk 2 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:57]: `streamer_settings` is preferable, agreed, wasn't sure about the naming at the time but didn't have a better idea. Will rename.

I intended to add the `stream_key` a bit later after setting up the streaming pipeline.

 `streamer_kv` does sound useful for things like `enable_clips` and similar toggles. Perhaps we could also store chat settings in this store for the time being?

Comment (bausano) [reply 001-MAIN_SCHEMA.md:57]: I thought that using a single `user` table which is in charge of auth and then extending it with `streamers` that fk's into it would simplify bunch of stuff.

Going to rethink now the semantics around the settings tables so that we don't start with something that is messy/prone to mess (e.g. by not knowing which table to query for particular settings etc.)

File: 001-MAIN_SCHEMA.md (hunk 3 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:136]:
```
- `last_subscribed_at` (timestamp)
  - when the user last subscribed
- `last_charged_at` (timestamp)
  - when the user was last charged
```

Comment (bausano) [reply 001-MAIN_SCHEMA.md:136]: Would be good to know when to charge them again :kiwi_fruit:

File: 001-MAIN_SCHEMA.md (hunk 4 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:42]: I thought to avoid DMs for the time being as it's complication and there's a plethora of other services to do DM's on. That was the rationale here, allow users to link their social if they wish to, thereby allowing connection.

File: 001-MAIN_SCHEMA.md (hunk 5 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:137]: You're right, we discussed it and as of yet I don't have a good idea of how to structure it, hence I didn't include it. Should've mentioned that. Another related topic is to research 3rd party apps such as streamlabs or streamelements.

`activity_log` table with a type enum and some text/json field for additional unstructured info is a good idea. These two columns + `streamer_id`?

File: 001-MAIN_SCHEMA.md (hunk 6 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:48]: Perhaps we don't store this in db at all, instead maintain a json file in s3 for each streamer. Would be more flexible when streamers come to us with custom requests I suppose.

Alternatively, we might want to render the json setting server side and serve HTML but that I assume has the big con of not leveraging FE react rendering with which we could do more interesting stuff easier if we just loaded something more sophisticated?

Comment (bausano) [reply 001-MAIN_SCHEMA.md:48]: Is there a structure you have in mind based on `about`s you've seen around?

File: 001-MAIN_SCHEMA.md (hunk 2 above)
Comment (kouks) [reply 001-MAIN_SCHEMA.md:57]: I agree that having 2 separate endpoints for fetching user data (settings, account info etc), and one for fetching streamers (followed streamers, recommended streamers, info about a stream) is better than trying to merge both into `/api/users` or wherever. This could however be handle purely by the backend.

File: 001-MAIN_SCHEMA.md (hunk 4 above)
Comment (kouks) [reply 001-MAIN_SCHEMA.md:42]: Understandable... honestly no one but spam bots use twitch DMs anyway

File: 001-MAIN_SCHEMA.md (hunk 6 above)
Comment (kouks) [reply 001-MAIN_SCHEMA.md:48]: what I'd have in mind is an object with a couple fields like so

`title: string`
`banner: { url, blurhash }`
`body: string`

where `body` could allow MD. I think a majority of people that use streaming services are familiar with basic markdown like

- lists
- **bold**
- _italic_

I could then either use a library to render this or even do a custom, limited implementation with these 3 basic features and maybe links.

File: 001-MAIN_SCHEMA.md (hunk 5 above)
Comment (kouks) [reply 001-MAIN_SCHEMA.md:137]: yeah `streamer_id` or `user_id` guess we could have 2 separate tables, one for account settings activity, one for stream activity, some `action` enum and some `data` json sounds like it's more than enough

File: 001-MAIN_SCHEMA.md (hunk 1 above)
Comment (kouks) [reply 001-MAIN_SCHEMA.md:9]: twitch lets you choose email or sms, sends a text with a 7 digit code on login (which is imo the simplest way), I'd probably avoid scanning QR code and using google authenticator

File: 001-MAIN_SCHEMA.md (hunk 6 above)
Comment (bausano) [reply 001-MAIN_SCHEMA.md:48]: How about the 3-columns orientation, apparently that is quite important to retain.

Comment (kouks) [reply 001-MAIN_SCHEMA.md:48]: I'd say make it an array of those objects. We can still keep it as 3 columns on the frontend, the number of banners can be variable.

File: 001-MAIN_SCHEMA.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+For large images we store both `url` and [`blurhash`][blurhash] strings.
+[Blurhash][blurhash] generates 20-30 character strings that are used to display a blurred version of the image while the actual image is loading.
+
+# `users`
+
+- `primary_email` (text, index)
+- `primary_email_confirmation_code` (smallint, optional)
+  - 5 digits
+  - is confirmed when this is null
+- `salt` (char 16)
```
Comment (kouks) [inline 001-MAIN_SCHEMA.md:15]: what exactly is this salt for?

Comment (kouks) [reply 001-MAIN_SCHEMA.md:15]: if it's for adding salt to the password I'm pretty sure this shouldn't be stored on the db

Comment (bausano) [reply 001-MAIN_SCHEMA.md:15]: > The salt can and should be stored right next to the salted and hashed password. Additionally, the salt should be unique per password. (45 upvotes)

> TL;DR - You can store the salt in plaintext without any form of obfuscation or encryption, but don't just give it out to anyone who wants it. (855 upvotes)

From https://security.stackexchange.com/questions/17421/how-to-store-salt

I am removing this column as the `bcrypt` dependency I use creates a string which contains the hash+salt compactly together. I am retaining just a single column which contains this compact string. However, to my knowledge, you should store a unique salt for each user with the hashed password.

Comment (kouks) [reply 001-MAIN_SCHEMA.md:15]: I thought the idea was that if database was leaked, the attackers couldn't bruteforce against all the hashed passwords in the database. But you probably have it researched better.

Comment (bausano) [reply 001-MAIN_SCHEMA.md:15]: Unique salt for each user is important because that prevents a scenario where the attacker takes a database of common passwords and rehashes it with a single salt. This way, the attacker has to rehash their database for _each_ user's salt.

Now, you could store the salt in another database from your main one, but I am assuming the thinking goes something like "if you managed to get your main database leaked, having a second database that's not leaked is improbable".
